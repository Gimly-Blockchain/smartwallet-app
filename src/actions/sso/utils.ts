import { IssuerPublicProfileSummary } from './types'
import { Identity } from 'jolocom-lib/js/identity/identity'

export const parsePublicProfile = (identity: Identity) => {
  const { publicProfile, did } = identity
  if (!publicProfile) {
    return {
      did,
    }
  }
  const { id, ...parsedProfile } = publicProfile.claim
  return {
    did,
    publicProfile: parsedProfile as IssuerPublicProfileSummary,
  }
}
