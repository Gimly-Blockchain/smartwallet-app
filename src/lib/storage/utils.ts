// TODO rework
// Multiple credentials of the same type, i.e. ['Credential', 'ProofOfNameCredential']
import {
  CredentialEntity,
  VerifiableCredentialEntity,
} from './entities'

interface ModifiedCredentialEntity {
  propertyName: string
  propertyValue: string[]
  verifiableCredential: VerifiableCredentialEntity
}

export const groupAttributesByCredentialId = (
  credentials: CredentialEntity[],
): ModifiedCredentialEntity[] => {
  // Convert values to arrays for easier concatination later
  const modifiedAttributes = credentials.map(credential => ({
    ...credential,
    propertyValue: [credential.propertyValue], // "eugeniu@jolocom.com" -> ["eugeniu@jolocom.com"]
  }))

  // Helper function
  const findByCredId = (
    arrToSearch: ModifiedCredentialEntity[],
    value: ModifiedCredentialEntity,
  ) =>
    arrToSearch.findIndex(
      entry => entry.verifiableCredential.id === value.verifiableCredential.id,
    )

  debugger
  return modifiedAttributes.reduce(
    (acc: ModifiedCredentialEntity[], curr: ModifiedCredentialEntity) => {
      const matchingIndex = findByCredId(acc, curr)

      // This seems to be a hack, uniting individual claims from the credential back together
      if (matchingIndex >= 0) {
        acc[matchingIndex].propertyValue = [
          ...acc[matchingIndex].propertyValue,
          ...curr.propertyValue,
        ]
        return acc
      } else {
        return [...acc, curr]
      }
    },
    [],
  )
}
