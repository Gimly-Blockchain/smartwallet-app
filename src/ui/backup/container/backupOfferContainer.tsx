import React from 'react'
import { connect } from 'react-redux'
import { ThunkDispatch } from '../../../store'
import { manualBackup, setAutoBackup } from '../../../actions/recovery'
import { NavigationScreenProps } from 'react-navigation'
import BackupOfferComponent from '../components/backupOfferComponent'
import { navigationActions } from '../../../actions'
import { routeList } from '../../../routeList'
import { withLoading } from '../../../actions/modifiers'

interface Props
  extends ReturnType<typeof mapDispatchToProps>,
    ReturnType<typeof mapStateToProps>,
    NavigationScreenProps {}

export class BackupOfferContainer extends React.Component<Props> {
  private manualBackup = async () => {
    await this.props.manualBackup()
    this.props.close()
  }
  public render() {
    const { close, enableAutoBackup } = this.props
    return (
      <BackupOfferComponent
        close={close}
        enableAutoBackup={enableAutoBackup}
        manualBackup={this.manualBackup}
      />
    )
  }
}

const mapStateToProps = () => ({})
const mapDispatchToProps = (dispatch: ThunkDispatch) => {
  return {
    enableAutoBackup: async () => {
      await dispatch(withLoading(setAutoBackup(true)))
      dispatch(navigationActions.navigate({ routeName: routeList.Settings }))
    },
    manualBackup: async () => await dispatch(manualBackup()),
    close: () =>
      dispatch(navigationActions.navigate({ routeName: routeList.Settings })),
  }
}

export const BackupOffer = connect(
  mapStateToProps,
  mapDispatchToProps,
)(BackupOfferContainer)
