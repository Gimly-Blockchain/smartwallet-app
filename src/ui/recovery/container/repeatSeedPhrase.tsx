import * as React from 'react'
import RepeatSeedPhraseComponent from '../components/repeatSeedPhrase'
import { ThunkDispatch } from '../../../store'
import { navigationActions, recoveryActions } from '../../../actions'
import { connect } from 'react-redux'
import { withLoading } from '../../../actions/modifiers'
import { toggleLoading } from '../../../actions/account'

interface Props extends ReturnType<typeof mapDispatchToProps> {
  navigation: { state: { params: any } } // TODO Type?
}

interface State {
  note: string
  sorting: {}
  randomWords: string[]
}

class RepeatSeedPhraseContainer extends React.Component<Props, State> {
  public state = {
    note:
      'If you have noted down your phrase, put the six given words on their right places.',
    sorting: {},
    randomWords: [],
  }

  private selectPosition = (id: number) => {
    const { sorting, randomWords } = this.state
    if (sorting[id]) {
      randomWords.unshift(sorting[id] as never)
      delete sorting[id]
    } else {
      const currentWord: string = randomWords[0]
      randomWords.shift()
      sorting[id] = currentWord
    }

    this.setState({
      sorting,
      randomWords,
    })
  }

  private checkMnemonic = () => {
    const { sorting } = this.state
    const { mnemonic } = this.props.navigation.state.params
    let isCorrect = true
    const mnemonicArray = mnemonic.split(' ')
    Object.keys(sorting).forEach(key => {
      if (mnemonicArray[key] !== sorting[key]) {
        isCorrect = false
      }
    })

    if (isCorrect) {
      this.props.setSeedPhraseSaved()
    } else {
      this.setState({
        sorting: {},
        randomWords: Object.keys(sorting).map(key => sorting[key]),
        note:
          'The order was not correct. Try again with another six words from your secure phrase.',
      })
    }
  }

  public componentDidMount(): void {
    const randomWords = []
    const { mnemonic } = this.props.navigation.state.params
    const mnemonicArray = mnemonic.split(' ')
    while (randomWords.length < 6) {
      const random = Math.floor(Math.random() * mnemonicArray.length)
      if (randomWords.indexOf(mnemonicArray[random]) === -1) {
        randomWords.push(mnemonicArray[random])
      }
    }
    this.setState({ randomWords })
  }

  public render(): JSX.Element {
    const { back } = this.props
    const { note, sorting, randomWords } = this.state
    return (
      <RepeatSeedPhraseComponent
        note={note}
        mnemonicSorting={sorting}
        randomWords={randomWords}
        back={back}
        checkMnemonic={this.checkMnemonic}
        selectPosition={this.selectPosition}
      />
    )
  }
}

const mapDispatchToProps = (dispatch: ThunkDispatch) => ({
  back: () => dispatch(navigationActions.goBack),
  setSeedPhraseSaved: () =>
    dispatch(withLoading(toggleLoading)(recoveryActions.setSeedPhraseSaved())),
})

export const RepeatSeedPhrase = connect(
  null,
  mapDispatchToProps,
)(RepeatSeedPhraseContainer)
