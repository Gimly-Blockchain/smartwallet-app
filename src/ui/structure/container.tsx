import React, { ReactNode } from 'react'
import { StyleSheet, ViewStyle, StyleProp, SafeAreaView } from 'react-native'

/**
 * Container
 *
 * The main use for this component is to have a full height and width component
 * that can be used in different components. It should be minimal in its style
 * so as to be reusable without constant overriding of its defaults.
 */

const styles = StyleSheet.create({
  container: {
    justifyContent: 'center',
    alignItems: 'center',
    height: '100%',
    width: '100%',
  },
})

interface Props {
  testID?: string
  children: ReactNode
  style?: StyleProp<ViewStyle>
}

export const Container: React.FC<Props> = props => (
  <SafeAreaView testID={props.testID} style={[styles.container, props.style]}>
    {props.children}
  </SafeAreaView>
)
