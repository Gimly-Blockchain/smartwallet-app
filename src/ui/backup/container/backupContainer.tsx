import React from 'react'
import { connect } from 'react-redux'
import { ThunkDispatch } from '../../../store'
import BackupComponent from '../components/backupComponent'
import { RootState } from '../../../reducers'
import {
  disableAndRemoveBackup,
  manualBackup,
  setAutoBackup,
} from '../../../actions/recovery'
import { timeout } from '../../../utils/asyncTimeout'
import { NavigationScreenProps } from 'react-navigation'
import { Alert } from 'react-native'

interface Props
  extends ReturnType<typeof mapDispatchToProps>,
    ReturnType<typeof mapStateToProps>,
    NavigationScreenProps {}

interface State {
  isLoading: boolean
}

export class BackupContainer extends React.Component<Props, State> {
  state = {
    isLoading: false,
  }

  private toggleLoading = () => {
    this.setState({ isLoading: !this.state.isLoading })
  }

  private enableAutoBackup = async () => {
    this.toggleLoading()
    await timeout(500)
    try {
      await this.props.enableAutoBackup()
    } catch (e) {
      // TODO show notification
      console.warn(e)
    } finally {
      this.toggleLoading()
    }
  }

  private onDisableAutoBackup = async () => {
    Alert.alert(
      'Are you sure?',
      'By disabling this function you will erase all information from our server',
      [
        { text: 'Cancel', style: 'cancel' },
        { text: 'OK', style: 'default', onPress: this.disableAutoBackup },
      ],
    )
  }

  private disableAutoBackup = async () => {
    this.toggleLoading()
    await timeout(500)
    try {
      await this.props.disableAutoBackupAndDelete()
    } catch (e) {
      // TODO show notification
      console.warn(e)
    } finally {
      this.toggleLoading()
    }
  }

  private exportBackup = async () => {
    this.toggleLoading()
    try {
      await this.props.exportBackup()
    } catch (e) {
      console.warn(e)
    } finally {
      this.toggleLoading()
    }
  }

  public render() {
    const { isAutoBackupEnabled, lastBackup, navigation } = this.props
    const { isLoading } = this.state
    return (
      <BackupComponent
        isLoading={isLoading}
        onDisableAutoBackup={this.onDisableAutoBackup}
        isAutoBackupEnabled={isAutoBackupEnabled}
        exportBackup={this.exportBackup}
        enableAutoBackup={this.enableAutoBackup}
        goBack={() => navigation.pop()}
        lastBackup={lastBackup}
      />
    )
  }
}

const mapStateToProps = (state: RootState) => ({
  isAutoBackupEnabled: state.settings.autoBackup,
  lastBackup: state.settings.lastBackup,
})
const mapDispatchToProps = (dispatch: ThunkDispatch) => ({
  enableAutoBackup: async () => await dispatch(setAutoBackup(true)),
  disableAutoBackupAndDelete: () => dispatch(disableAndRemoveBackup()),
  exportBackup: () => dispatch(manualBackup()),
})

export const Backup = connect(
  mapStateToProps,
  mapDispatchToProps,
)(BackupContainer)
