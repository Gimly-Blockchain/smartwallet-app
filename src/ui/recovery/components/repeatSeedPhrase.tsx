import * as React from 'react'
import { StyleSheet, Text, TouchableOpacity, View } from 'react-native'
import { Container } from '../../structure'
import { JolocomTheme } from '../../../styles/jolocom-theme'
import { Button } from 'react-native-material-ui'

const styles = StyleSheet.create({
  container: {
    backgroundColor: JolocomTheme.primaryColorBlack,
  },
  noteSection: {
    marginTop: 20,
    flex: 0.8,
    justifyContent: 'center',
  },
  note: {
    textAlign: 'center',
    lineHeight: 26,
    color: JolocomTheme.primaryColorSand,
    fontSize: JolocomTheme.labelFontSize,
    fontFamily: JolocomTheme.contentFontFamily,
  },
  mnemonicContainer: {
    flex: 1,
    flexDirection: 'row',
  },
  mnemonic: {
    margin: 8,
    color: JolocomTheme.primaryColorWhite,
  },
  selector: {
    width: 109,
    height: 32,
    borderRadius: 4,
    borderStyle: 'solid',
    borderWidth: 1,
    borderColor: '#942f51',
    margin: 10,
  },
  buttonSection: {
    marginTop: 'auto',
  },
  buttonContainer: {
    borderRadius: 4,
    height: 48,
    paddingHorizontal: 25,
    backgroundColor: JolocomTheme.primaryColorPurple,
  },
  buttonText: {
    paddingVertical: 15,
    fontWeight: '100',
    fontSize: JolocomTheme.headerFontSize,
    color: JolocomTheme.primaryColorWhite,
    fontFamily: JolocomTheme.contentFontFamily,
  },
})

interface RepeatSeedPhraseProps {
  note: string
  mnemonicSorting: {}
  randomWords: string[]
  back: () => void
  checkMnemonic: () => void
  selectPosition: (id: number) => void
}
const RepeatSeedPhraseComponent = ({
  note,
  mnemonicSorting,
  randomWords,
  back,
  checkMnemonic,
  selectPosition,
}: RepeatSeedPhraseProps) => (
  <Container style={styles.container}>
    <View style={styles.noteSection}>
      <Text style={styles.note}>{note}</Text>
    </View>
    <View style={styles.mnemonicContainer}>
      {randomWords.map(key => (
        <Text style={styles.mnemonic}>{key}</Text>
      ))}
    </View>
    <View style={{ display: 'flex', flexDirection: 'row' }}>
      <View>
        {new Array(6).fill('').map((e, i) => (
          <TouchableOpacity
            key={i}
            style={styles.selector}
            onPress={() => selectPosition(i)}
          >
            <Text style={styles.note}>
              {mnemonicSorting[i] ? mnemonicSorting[i] : i + 1}
            </Text>
          </TouchableOpacity>
        ))}
      </View>
      <View>
        {new Array(6).fill('').map((e, i) => (
          <TouchableOpacity
            key={i}
            style={styles.selector}
            onPress={() => selectPosition(i + 6)}
          >
            <Text style={styles.note}>
              {mnemonicSorting[i + 6] ? mnemonicSorting[i + 6] : i + 7}
            </Text>
          </TouchableOpacity>
        ))}
      </View>
    </View>
    <View style={styles.buttonSection}>
      <Button
        style={{ container: styles.buttonContainer, text: styles.buttonText }}
        onPress={randomWords.length ? back : checkMnemonic}
        raised
        upperCase={false}
        text={randomWords.length ? 'Show my phrase again' : 'Confirm and check'}
      />
    </View>
  </Container>
)

export default RepeatSeedPhraseComponent
