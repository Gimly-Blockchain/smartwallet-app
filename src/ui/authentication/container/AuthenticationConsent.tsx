import React from 'react'
import { connect } from 'react-redux'
import { AuthenticationConsentComponent } from '../components/AuthenticationConsent'
import { StateAuthenticationRequestSummary } from 'src/reducers/sso'
import { RootState } from 'src/reducers'
import { cancelSSO } from 'src/actions/sso'
import { sendAuthenticationResponse } from 'src/actions/sso/authenticationRequest'
import {ThunkDispatch} from '../../../store'

interface ConnectProps {}

interface Props extends ConnectProps {
  activeAuthenticationRequest: StateAuthenticationRequestSummary
  confirmAuthenticationRequest: () => typeof sendAuthenticationResponse
  cancelAuthenticationRequest: () => typeof cancelSSO
}

interface State {}

export class AuthenticationConsentContainer extends React.Component<
  Props,
  State
> {
  render() {
    return (
      <AuthenticationConsentComponent
        activeAuthenticationRequest={this.props.activeAuthenticationRequest}
        confirmAuthenticationRequest={this.props.confirmAuthenticationRequest}
        cancelAuthenticationRequest={this.props.cancelAuthenticationRequest}
      />
    )
  }
}

const mapStateToProps = (state: RootState) => ({
  activeAuthenticationRequest: state.sso.activeAuthenticationRequest,
})

const mapDispatchToProps = (dispatch: ThunkDispatch) => ({
  confirmAuthenticationRequest: () => sendAuthenticationResponse,
  cancelAuthenticationRequest: () => cancelSSO,
})

export const AuthenticationConsent = connect(
  mapStateToProps,
  mapDispatchToProps,
)(AuthenticationConsentContainer)
