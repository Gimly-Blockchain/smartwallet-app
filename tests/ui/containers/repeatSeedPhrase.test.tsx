import React from 'react'
import { shallow } from 'enzyme'
import { createMockNavigationScreenProp } from 'tests/utils'
import { RepeatSeedPhraseContainer } from '../../../src/ui/recovery/container/repeatSeedPhrase'

const mockSeedPhrase = 'mock seed Phrase of at least six words or more'
describe('repeatedPhrase container', () => {
  const setSeedPhraseSaved = jest.fn()
  const props: RepeatSeedPhraseContainer['props'] = {
    openSeedPhrase: jest.fn(),
    setSeedPhraseSaved,
    navigation: createMockNavigationScreenProp({
      state: {
        params: {
          mnemonic: mockSeedPhrase,
        },
      },
    }),
  }

  it('mounts correctly and matches snapshot', () => {
    RepeatSeedPhraseContainer.prototype.getRandomWords = jest.fn(() =>
      mockSeedPhrase.split(' ').slice(0, 6),
    )
    const rendered = shallow(<RepeatSeedPhraseContainer {...props} />)
    expect(rendered).toMatchSnapshot()
  })
  describe('component methods', () => {
    let instance: any
    let wrapper: any
    beforeEach(() => {
      RepeatSeedPhraseContainer.prototype.getRandomWords = jest.fn(() =>
        mockSeedPhrase.split(' ').slice(0, 6),
      )
      const rendered = shallow(<RepeatSeedPhraseContainer {...props} />)
      instance = rendered.instance()
      wrapper = rendered
      setSeedPhraseSaved.mockClear()
    })

    it('should selectPosition correctly', () => {
      expect(wrapper.state('sorting')).toEqual({})
      expect(wrapper.state('randomWords')).toContain('mock')
      instance.selectPosition(1)
      expect(wrapper.state('sorting')).toEqual({ 1: 'mock' })
      expect(wrapper.state('randomWords')).not.toContain('mock')
    })

    it('should correctly validate Mnemonic', () => {
      wrapper.setState({
        sorting: {
          0: 'mock',
          1: 'seed',
          2: 'Phrase',
          3: 'of',
          4: 'at',
          5: 'least',
        },
      })
      instance.checkMnemonic()
      expect(setSeedPhraseSaved).toBeCalled()
    })
    it('should not validate wrong sorting', () => {
      wrapper.setState({
        sorting: {
          0: 'seed',
          1: 'Phrase',
          2: 'at',
          3: 'mock',
          4: 'of',
          5: 'least',
        },
      })
      instance.checkMnemonic()
      expect(setSeedPhraseSaved).not.toBeCalled()
      expect(wrapper.state('sorting')).toEqual({})
      expect(wrapper.state('randomWords').length).toEqual(6)
    })
  })
})
