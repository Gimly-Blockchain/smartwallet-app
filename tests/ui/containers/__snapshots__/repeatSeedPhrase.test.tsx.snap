// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`repeatedPhrase container mounts correctly and matches snapshot 1`] = `
<RepeatSeedPhraseComponent
  back={[Function]}
  checkMnemonic={[Function]}
  mnemonicSorting={Object {}}
  note="If you have noted down your phrase, put the six given words on their right places."
  randomWords={
    Array [
      "mock",
      "seed",
      "Phrase",
      "of",
      "at",
      "least",
    ]
  }
  selectPosition={[Function]}
/>
`;
